### PR TITLE
docs: draft: Intro to Hardened JS

### PR DESCRIPTION
**Pull request by dckc**
todo:
 - [ ] publish the 8 minute `makeMint` video segment on its own
 - [ ] zero-install dev env #554
 - [ ] turn into slides with remark?
   - https://hackmd.io/@ZFlbCaR1Q06UzHk6tcn2dg/rk1yMNeQF#/5
 - [ ] remove brainstorm appendix
 - [ ] get team decision on whether this belongs on master, and if so, is this the right directory?

### CODE REVIEW

File: main/platform/hardened-js.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+3
+> c2.increment();
+13
+> [c1.increment(), c2.increment()];
+[ 4, 14 ]
+```
+
+ - An object is a record of functions that close over shared state.
+ - A `makeCounter` function take the place of `class Counter`.
+ - `this` is not in Jessie.
```
Comment (michaelfig) [inline main/platform/hardened-js.md:73]:
```suggestion
 - The `this` keyword is not part of Jessie, so neither are constructors.
 - A `makeCounter` function takes the place of the `class Counter` constructor syntax.
```

File: main/platform/hardened-js.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+> c3.increment = () => { console.log('launch the missiles!'); }
+TypeError: Cannot assign to read only property 'increment' of object '#<Object>'
+> c3.increment()
+15
+```
+
+ - _caveat_: exception is thrown in strict mode. REPL might not throw.
+ - regardless, the object defended itself
+ - Jessie pre-defines `harden` (_as does Agoric smart contract framework_)
+   - `npm install @agoric/harden` and `import harden from '@agoric/harden';` otherwise
```
Comment (michaelfig) [inline main/platform/hardened-js.md:115]: We really don't want people to use `@agoric/harden`, since ses is ready for this usage pattern.

```suggestion
   - see the `ses` [hardened Javascript package](https://github.com/endojs/endo/tree/master/packages/ses#harden) otherwise
```

File: main/platform/json-justin-data-expr.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+   - add, subtract, etc.: `+`, `-`, `*`, `/`, `%`
+   - parens `(x + 4) * y`
+   - comparison: `===`, `!==`, `<`, `>`, `<=`, `>=`
+       - **no `==`!**
+   - relational: `&&`, `||`
+   - bitwise: `&`, `|`, `^`
+   - shift: `<<`, `>>`
+     - unsigned: `>>>`
+   - runtime type check: `typeof thing == "string"`
+   - conditional: ``typeof thing == "string" ? thing : ${thing} ``
```
Comment (michaelfig) [inline main/platform/json-justin-data-expr.md:54]:
```suggestion
   - runtime type check: `typeof thing === "string"`
   - conditional (ternary operator): ``a > b ? a : b``
```

File: main/platform/json-justin-data-expr.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+ - `if ... then ... else`, `switch`
+ - `while`, `break`, `continue`
+ - `try` / `finally / `catch`
+ - `for (x of items) { }`
+   - **no `for (x in items)`!**
+
+_TODO: accessor methods?_
+
+
+## Destructing Assignments, Patterns, and Shorthand Properties
```
Comment (michaelfig) [inline main/platform/json-justin-data-expr.md:89]:
```suggestion
## Destructuring Assignments, Patterns, and Shorthand Properties
```

File: main/platform/json-justin-data-expr.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+```js
+const s = order.size;     // better as... 
+const { size: s } = order;
+
+const size = order.size; // better as...
+const { size } = order;
+
+const s1 = sizes[1];
+const s2 = sizes[2];
```
Comment (michaelfig) [inline main/platform/json-justin-data-expr.md:99]:
```suggestion
const s1 = sizes[0];
const s2 = sizes[1];
```

**Review by michaelfig — COMMENTED** [review]
Here's a first pass.   Looking pretty good!

File: main/platform/json-justin-data-expr.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+   - `1, 2, 3, true, false, "abc", null`
+   - `"emoji: \uD83D\uDC36"`, `"🐶"`
+ - Array: `[1, true, "three"]`
+ - Record (aka Object): `{ "size": 5, "color": "blue" }`
+
+
+## Justin: Safe Expressions
+
+ - more flexible syntax:
+   - string delimters: `"abc"` or  `'abc'`
```
Comment (michaelfig) [inline main/platform/json-justin-data-expr.md:14]:
```suggestion
   - string delimiters: `"abc"` or  `'abc'`
```

File: main/platform/json-justin-data-expr.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+   - trailing commas: `{ size: 1, }`, `[1, 2, 3,]`
+   - un-quoted properties: `{ size: 1, color: "blue" }`
```
Comment (michaelfig) [inline main/platform/json-justin-data-expr.md:16]:
```suggestion
   - un-quoted property names: `{ size: 1, color: "blue" }`
```

File: main/platform/hardened-js.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
+# Intro to Hardened JavaScript
+
+_Status: WIP_.
+
```
Comment (michaelfig) [inline main/platform/hardened-js.md:4]:
```suggestion

TL;DR: Here are the steps to play whack-a-mole with the Jessie linter:

1. If not already configured, run `yarn add eslint @jessie.js/eslint-plugin`
2. If not already configured, add the following to your `package.json`:

```json
  "eslintConfig": {
    "extends": [
      "@jessie.js"
    ]
  }
```// TODO: remove this comment to complete the json fence
3. Put `// @jessie-check` at the beginning of your `.js` source file.
4. Run `yarn eslint --fix path/to/your-source.js`
5. Follow the linter's advice to edit your file, then go back to step 4.
```

**Review by michaelfig — COMMENTED** [review]
Here's some more edits.

**Comment by dckc** [comment]
Integrating slides looks more invasive than I would like.

In my searching, all roads lead to [Presentation \| vuepress\-theme\-hope](https://vuepress-theme-hope.github.io/guide/markdown/presentation/). Adding a whole theme seems risky.

**Comment by dckc** [comment]
## use makeMint as focal point of Hardened JS

@michaelfig @katelynsills @erights I think I figured out how to fit this material with the rest: build up to `makeMint`, at which point readers are qualified to study ERTP etc.

TODO: publish the 8 minute `makeMint` video segment on its own

**Comment by dckc** [comment]
@rowgraus this is the video I want to publish on its own:

**Watch**: [the mint pattern](https://youtu.be/iyuo0ymTt4g?t=1525),
an 8 minute segment starting at 25:00 in

 - [Agoric \+ Protocol Labs // Higher\-order Smart Contracts across Chains \- Mark Miller \- YouTube](https://www.youtube.com/watch?v=iyuo0ymTt4g).

It's the punch-line for these slides now: https://github.com/Agoric/documentation/blob/js-hardened-intro/main/platform/hardened-js.md#electronic-rights-mint-and-purse

the rest is ordinary programming that's not that special to Agoric or OCaps:

https://github.com/Agoric/documentation/blob/js-hardened-intro/main/platform/json-justin-data-expr.md

**Comment by erights** [comment]
Thanks! Just watched and I agree that those 8 minutes are coherent and powerful by themselves. Good choice!

**Comment by dckc** [comment]
fun with hackmd slides: https://hackmd.io/jHfqwBhnRteUIk_bI27wvQ?both

**Comment by dckc** [comment]
fun with code visualization: [counter example on JavaScript Tutor](https://pythontutor.com/javascript.html#code=const%20makeCounter%20%3D%20init%20%3D%3E%20%7B%0A%20%20let%20value%20%3D%20init%3B%0A%20%20return%20%7B%0A%20%20%20%20increment%3A%20%28%29%20%3D%3E%20%28value%20%2B%3D%201%29,%0A%20%20%20%20decrement%3A%20%28%29%20%3D%3E%20%28value%20-%3D%201%29,%0A%20%20%20%20makeOffsetCounter%3A%20delta%20%3D%3E%20makeCounter%28value%20%2B%20delta%29,%0A%20%20%7D%3B%0A%7D%3B%0A%0Aconst%20c1%20%3D%20makeCounter%281%29%3B%0Aconsole.log%28c1.increment%28%29%29%3B%0A%0Aconst%20c2%20%3D%20c1.makeOffsetCounter%2810%29%3B%0Aconsole.log%28c2.increment%28%29%29%3B%0Aconsole.log%28%5Bc1.increment%28%29,%20c2.increment%28%29%5D%29%3B&curInstr=26&heapPrimitives=nevernest&mode=display&origin=opt-frontend.js&py=js&rawInputLstJSON=%5B%5D)

 - [ ] makeMint to appear

**Comment by dckc** [comment]
overtaken by #579
